### PR TITLE
cythonize protocol.binary

### LINKED ISSUE #13: cythonize binary protocol
I haven't done any profiling on this yet but I think it's safe to say that binary protocol implementation will be a significant portion of the cost. It's worth optimizing that as much as possible. We should cythonize the binary protocol implementation and if necessary, also add type annotations/cdefs.

### PR DESCRIPTION
**Pull request by blampe**
Fixes #13.

I've got a benchmark that runs `dumps` on a struct with some simple nesting and a 100-element list. Without cython it's about 0.014s per `dumps` call. With this and #44 it's reduced to ~0.005s. Interestingly `to_primitive` only takes 0.0007s on the same structure, probably because it's not converting it to an intermediate wire value?

@abhinav @breerly @junchaowu

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -2,6 +2,22 @@
...
+    if tuple(int(v) for v in Cython.__version__.split('.')) > (0, 17, 0):
+        extra = {
+            'ext_modules': Cython.Build.cythonize([
+                'thriftrw/protocol/binary.py',
+                'thriftrw/wire/value.py',
+                'thriftrw/spec/*.py',
+            ]),
+        }
+except ImportError:
+    pass
```
Comment (abhinav) [inline setup.py]: I think what you want to do instead is rename the files you want to compile to `.pyx` and always compile them to C if cython is present. If it's not present, we should assume that the C files are present (they will need to be added to MANIFEST.in), and add them to `ext_modules` regardless. See http://stackoverflow.com/a/4515279/58536.

That way users don't have to depend on cython. We'll only need it during development and release.

File: tox.ini [hunk 2]
```diff
@@ -1,8 +1,11 @@
 [tox]
-envlist = py27,py33,py34,py35,pypy,pypy3,cover,flake8,docs
+envlist = py27,py33,py34,py35,pypy,pypy3,cython,cython3,cover,flake8,docs
+usedevelop = true
 
 [testenv]
-commands = py.test -svv tests
+commands =
+    - make clean
+    py.test -svv tests
```
Comment (abhinav) [inline tox.ini]: Tox manages its own environment. `make clean` won't modify tox's environment.

File: setup.py [hunk 3]
```diff
@@ -1,6 +1,45 @@
...
+# compile the packaged C files.
+try:
+    import Cython.Distutils
+    # cython didn't support "from __future__" statements until 0.17.
+    if tuple(int(v) for v in Cython.__version__.split('.')) > (0, 17, 0):
+        extension_filetype = '.pyx'
+        cmdclass.update(build_ext=Cython.Distutils.build_ext)
+except ImportError:
+    extension_filetype = '.c'
+
```
Comment (abhinav) [inline setup.py]: Personally, I would list the cythonized modules explicitly. Avoids confusion or unexpected behavior later. But I don't feel strongly about this.

File: tox.ini [hunk 4]
```diff
@@ -1,8 +1,11 @@
 [tox]
 envlist = py27,py33,py34,py35,pypy,pypy3,cover,flake8,docs
+usedevelop = true
 
 [testenv]
-commands = py.test -svv tests
+commands =
+    python setup.py clean --all build_ext --force --inplace
```
Comment (abhinav) [inline tox.ini]: need to add this to `cover` too, right?

**Comment by abhinav** [comment]
:+1:

**Comment by blampe** [comment]
Merged into 0.5.0 branch.

**Pull request closed without merging**